docs/http.md

```diff
@@ -0,0 +1,298 @@
+headers, options, and allow the caller to stream the request and response body.
+
+#### Connection Pool
+
+Maintains the pool of connections, keeping track of when idle connections
+expire. The `with_connection` method takes a block, which ensures borrowed
+connections are always returned to the pool.
+
+#### Session
+
+Has methods for selecting a route based on the client's configuration. If a
+route is selected, but the initial request made using that route fails, then the
+session will select another route, the request will be retried. This process
+repeats until a request succeeds or all routes are exhausted.
+
+Once a route has been successfully used once, it will be reused for the duration
+of the session. For example, if the node request succeeds, then the agent will
+use the same server for the catalog and pluginsync requests.
+
+#### Services
+
+Routes have logical names based on the type of service provided, such as `:ca`.
+Puppet supports four services currently: `:ca` and `:report` handle certs and
+reports. The `:fileserver` service handles puppet file metadata and content
+requests, such as pluginsync and file resources with `source => 'puppet://'`.
+The `:puppet` service handles nodes, facts and catalogs, and is also the
+fallback for the other three services.
+
+#### Route
+
+Represents the route to a service. The route contains the destination server's
+host and port, and has methods that are appropriate for that route. For example,
+the `CaRoute` has a `get_certificate` method.
+
+The route handles serialization/deserialization of the entity, leveraging the existing
+`Puppet::Network::Format` objects.
+
+#### Resolvers
+
+Each resolver represents a different strategy for resolving the service name
+into a list of possible servers and ports to try. There are three resolvers:
+SRV, ServerList and DefaultResolver.
+
+### Routing
+
+The `Puppet::HTTP::Session` class will have a `route_to` method that takes a
+service name. The method will yield a `Puppet::HTTP::Route` specifying the
+resolved server host and port to use. For example:
+
+```ruby
+client = Puppet::HTTP::Client.new
+
+session = client.create_session
+session.route_to(:ca) do |route|
+  cert = route.get_certificate('ca')
+  puts "Retrieved cert #{cert.subject.to_utf8} from #{route.server}:#{route.port}"
+end
+```
+
+If an explicit server and port were specified on the command line or
+configuration, such as `puppet agent -t --server foo.example.com`, then the
+`route_to` method will always yield that value.
+
+Otherwise, the session will walk the list of resolvers in priority order:
+
+* DNS SRV
+* Server list
+* Default puppet settings
+
+If the `route_to` method yields a route, but it results in an exception, such as
+"connection refused", then the session will yield the next route, until all
+routes for all resolvers have been tried.
+
+If the caller successfully uses a route, then the session will yield the same
+route the next time `route_to` is called again.
```

Comment:
I assume you mean, that route will be the first tried, before following the priority list if that route cannot be reached?